### ISSUE TITLE
Regression in clang-format version 16?

### ISSUE THREAD

**Issue by prj-** [llvm/llvm-project#61498]
Steps to reproduce:
1. download [reproducer.tar.gz](https://github.com/llvm/llvm-project/files/11009158/reproducer.tar.gz) which includes `petscstring.h`, `.clang-format-15`, and `.clang-format-16`;
2. run `clang-format-15 --style=file:.clang-format-15 petscstring.h > 15.h`;
3. run `clang-format-16 --style=file:.clang-format-16 petscstring.h > 16.h`;
4. `diff 15.h 16.h`.

This should produce the following unwanted diff.
```diff
--- 15.h	2023-03-18 18:07:05
+++ 16.h	2023-03-18 18:07:02
@@ -474,7 +474,7 @@
 #if PetscHasBuiltin(__builtin_strchr)
   *c = (char *)__builtin_strchr(a, b);
 #else
-  *c = (char *)strchr(a, b);
+  *c   = (char *)strchr(a, b);
 #endif
   PetscFunctionReturn(PETSC_SUCCESS);
 }
```

For reference only.
```console
$ clang-format-16 -version
clang-format version 16.0.0
$ clang-format-15 -version
clang-format version 15.0.7
```

**Comment by rymiel** [comment]
Bisected to c0779756a0c4cc84d9f98714734d47879701cc3d / https://reviews.llvm.org/D134042

It seems it's being aligned to the similar assignment to `*tmp` of the following `PetscStrrchr()` function, which can be observed by making that variable longer; so definitely a bug

**Comment by rymiel** [comment]
Minimal reproducer:
```c++
#if x
a = 0;
#else
a   = 1;
#endif

#if x
abc = 0;
#else
abc = 1;
#endif
```

Minimal style:
```yaml
AlignConsecutiveAssignments: Consecutive
```

**Comment by prj-** [comment]
Do you have an idea on how we could bypass this? This is happening in four of our files, see https://gitlab.com/petsc/petsc/-/merge_requests/6203#note_1320939442.

**Comment by prj-** [comment]
Any update on this issue? This is in fact triggered in multiple places in our code, and the result is quite atrocious, see, e.g., this random line with extra spacing https://gitlab.com/petsc/petsc/-/blob/ab4ee011827fbbb77f789779c813f6db6bb0cbfa/src/sys/objects/kokkos/kinit.kokkos.cxx#L29.

**Comment by andli197** [comment]
I also wonder if there is any update to this issue? Sorry for the duplicate.

**Comment by owenca** [comment]
See https://reviews.llvm.org/D134042 for the discussion about reverting c077975 for now.

**Comment by prj-** [comment]
Thanks for the update.

**Comment by owenca** [comment]
> Minimal reproducer:
> 
> ```c++
> #if x
> a = 0;
> #else
> a   = 1;
> #endif
> 
> #if x
> abc = 0;
> #else
> abc = 1;
> #endif
> ```
> 
> Minimal style:
> 
> ```yaml
> AlignConsecutiveAssignments: Consecutive
> ```

clang-format 15 also fails on this one.

**Comment by owenca** [comment]
> Bisected to [c077975](https://github.com/llvm/llvm-project/commit/c0779756a0c4cc84d9f98714734d47879701cc3d) / https://reviews.llvm.org/D134042

Commit c077975 partially addressed a deficiency clang-format has had since who knows when. Although reverting it would take care of the "regression", it would not solve the problem in general. That is, clang-format would still merge all `#else`, `#elif`, `#elifdef`, and `#elifndef` blocks into a single block when doing whitespace replacements.

**Comment by owenca** [comment]
https://reviews.llvm.org/D150057

**Comment by Sedeniono** [comment]
I think a4c87f8ccaccc76fd7d1c6c2e639ca84b9ec7794 introduces a regression. I created #62892 for it.

**Comment by tstellar** [comment]
Since this change introduced a regression, I'm moving the to the 17.0.0 milestone. If you would like this included in the 16.0.x release, change the milestone back to 16.0.X and explain why it needs to be included.

**Comment by owenca** [comment]
The regression was fixed in 441108ccba94 and 8e85739a5f24.
